### DISCUSSION TITLE
Adaptive MCS on retransmission

### DISCUSSION THREAD

**Discussion by christsako** [srsran/srsRAN_Project#705] [General Help] [answered]
Hello to the community. I am trying to figure out the adaptation of MCS in the scheduler and for retransmissions I can see this:

// Compute the MCS and the number of PRBs, depending on the pending bytes to transmit.
    grant_prbs_mcs mcs_prbs =
        is_retx ? grant_prbs_mcs{h_ul.last_tx_params().mcs, h_ul.last_tx_params().rbs.type1().length()}
        
   So the MCS remains the same as the previous' failed initial transmission as well as the PRB allocation.  By viewing the spec I can see some MCS values (the ones over 29) reserved for retransmission. This isn't supported yet as a feature or do I miss sth in understanding?
   
   What is more, I see the harq process's TBS cannot change between initial transmission and retransmission. If the MCS could change and it was decided by the scheduler to give a smaller MCS due to current channel conditions, could that mean the number of PRBs should also change to guarantee the TBS stays the same? In general it isn't clear to me if MCS adaptation can be applied to retransmisison too.

Thanks in advance for any clarification on this.

**Comment by frankist** [comment, marked as answer]
The TBS cannot change across HARQ retransmissions, because the MAC PDU was already encoded at that stage, and to leverage the concepts of softbuffers and redundancy versions, the MAC PDU data needs to remain the same.

What you suggested about decreasing the MCS and increasing the number of PRBs, while keeping the TBS constant should be possible (otherwise, what would be the point of the DCI resending the chosen MCS in a retx?), but I didn't have time to look into it.
